docs/adr/0003-delivery-channels-templating.md

```diff
@@ -0,0 +1,88 @@
+# Templating engine for different notification types
+
+## Status
+
+Proposed
+
+## Context
+
+This ADR proposes a method on how to implement custom templating when a certain notification is sent through a communication channels (SMS, email, web push notifications etc).
+
+In a typical notification system, an object will be applied to a basic template to provide an end user with information targeting them. A lot of the basic design/layout of the template will be quite standard, but there will be places for values that will need to be inserted to customise it for the said end user. An example of this would be an birthday email notification, where a lot of the design and formatting of the email/subject will already be very well defined, and then insertion of custom data (eg. email id, name, etc ) is what will be needed by a templater.
+
+The content of this ADR will work off themes from the previous ADR (0002-delivery-channel-creation-and-integration.md).
+
+## Decision
+
+A notification having necessary information needs to be translated into a certain communication object (SMS, Email, etc) before it can be sent via the desired delivery channel (SendGrid, Twillio, Airship).
+
+An example of this would be a SMS, which needs the essential information to be available for it to be converted into an SMS object before being sent to the end user.
+
+```c#
+
+    public class SMS {
+        public string ToPhoneNumber { get; set; }
+        public string Body { get; set; }
+    }
+
+```
+
+Properties in the communication object will be populated through different sources based on the communication type.
+
+Eg. Properties in the above SMS object will be populated from the user preference store (azure table storage) from which the `ToPhoneNumber` will be consumed while the `Body` will be populated from the template store (azure table storage).
+
+### Creation of templates, how does someone create templates and in what language
+
+A template will be created based on the notification type, the communication type, and the communication object's property key.
+
+A few examples of this would be:
+
+```
+(Notification type) -> (Communication type) -> (Communication type property)
+
+New Lead            -> SMS                  -> Body
+
+New Lead            -> Email                -> Subject
+New Lead            -> Email                -> Body
+
+```
+
+Every communication type will have a similar communication object, ie, an SMS will use an SMSobject that has certain set of fields. Some of those fields would be required without which a notification would be impossible to be sent. Eg. A SMSObject should have `ToPhoneNumber` or EmailObject should have `ToEmail` and without these properties being populated a notification could not be sent. Furthermore, the `Body` for these 2 communication object would be equally important as there would be no point in sending an empty notification.
```

Comment:
Rather than `SMSObject`, `EmailObject`, can we use a more descriptive name?

e.g. `FormattedSmsMessage`, `FormattedEmailMessage`?

This better represents the fact that the object is the result of applying a template to a notification, the output of which is a message that's ready to be sent via a delivery channel.